apypie/foreman.py

```diff
@@ -0,0 +1,228 @@
+"""
+Apypie Foreman module
+
+opinionated helpers to use Apypie with Foreman
+"""
+import time
+
+from typing import cast, Optional, Set, Tuple
+
+from apypie.api import Api
+
+from apypie.resource import Resource  # pylint: disable=unused-import
+
+PER_PAGE = 2 << 31
+
+
+class ForemanApiException(Exception):
+    """
+    General Exception, raised by any issue in ForemanApi
+    """
+
+    def __init__(self, msg, error=None):
+        self.msg = msg
+        self.error = error
+        super().__init__()
+
+    def __repr__(self):
+        return str(self)
+
+    def __str__(self):
+        string = f'{self.__class__.__name__}: {self.msg}'
+        if self.error:
+            string += f' - {self.error}'
+        return string
+
+    @classmethod
+    def from_exception(cls, exc, msg):
+        """
+        Create a ForemanException from any other Exception
+
+        Especially useful to gather the error message from HTTP responses.
+        """
+        error = None
+        if hasattr(exc, 'response') and exc.response is not None:
+            try:
+                response = exc.response.json()
+                if 'error' in response:
+                    error = response['error']
+                else:
+                    error = response
+            except Exception:  # pylint: disable=broad-except
+                error = exc.response.text
+        return cls(msg=msg, error=error)
+
+
+class ForemanApi(Api):
+    """
+    apypie.Api with default settings and helper functions for Foreman
+    """
+
+    def __init__(self, **kwargs):
+        self.task_timeout = kwargs.pop('task_timeout', 60)
+        self.task_poll = 4
+        kwargs['api_version'] = 2
+        super().__init__(**kwargs)
+
+    def _resource(self, resource: str) -> 'Resource':
+        if resource not in self.resources:
+            raise ForemanApiException(msg="The server doesn't know about {0}, is the right plugin installed?".format(resource))
```

Comment:
```suggestion
            raise ForemanApiException(msg=f"The server doesn't know about {resource}, is the right plugin installed?")
```

Reply:
Oh yeah, that's what you get by copying 2.7 compatible code :)